2022Swerve/src/main/java/frc/robot/subsystems/SwerveSubsystem.java

```diff
@@ -0,0 +1,82 @@
+package frc.robot.subsystems;
+
+import com.kauailabs.navx.frc.AHRS;
+
+import edu.wpi.first.math.geometry.Rotation2d;
+import edu.wpi.first.math.kinematics.SwerveDriveKinematics;
+import edu.wpi.first.math.kinematics.SwerveModuleState;
+import edu.wpi.first.wpilibj.SPI;
+import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
+import edu.wpi.first.wpilibj2.command.InstantCommand;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import edu.wpi.first.wpilibj2.command.WaitUntilCommand;
+import frc.robot.Constants.DriveConstants;
+import frc.robot.hardware.AbsoluteEncoders;
+import frc.robot.hardware.MotorControllers;
+
+public class SwerveSubsystem extends SubsystemBase{
+    private final SwerveModule frontLeft = new SwerveModule(
+        MotorControllers.FrontLeftModuleDrive.getMotor(),
+        MotorControllers.FrontLeftModuleTurn.getMotor(),
+        AbsoluteEncoders.FrontLeftModule.getSupplier());
+
+    private final SwerveModule frontRight = new SwerveModule(
+        MotorControllers.FrontRightModuleDrive.getMotor(),
+        MotorControllers.FrontRightModuleTurn.getMotor(),
+        AbsoluteEncoders.FrontRightModule.getSupplier());
+
+    private final SwerveModule backLeft = new SwerveModule(
+        MotorControllers.BackLeftModuleDrive.getMotor(),
+        MotorControllers.BackLeftModuleTurn.getMotor(),
+        AbsoluteEncoders.BackLeftModule.getSupplier());
+
+    private final SwerveModule backRight = new SwerveModule(
+        MotorControllers.BackRightModuleDrive.getMotor(),
+        MotorControllers.BackRightModuleTurn.getMotor(),
+        AbsoluteEncoders.BackRightModule.getSupplier());
+
+    private AHRS gyro = new AHRS(SPI.Port.kMXP);
+
+    public SwerveSubsystem() {
+        new WaitUntilCommand(this::gyroReady)
+        .andThen(new InstantCommand(this::zeroHeading,this))
+        .schedule();
+    }
+
+    public boolean gyroReady() {
+        return !gyro.isCalibrating();
+    }
+
+    public void zeroHeading() {
+        gyro.reset();
+    }
+
+    public double getHeading() {
+        return Math.IEEEremainder(gyro.getAngle(), 360); //Clamps angle output between -180 and 180 degrees
+    }
+
+    public Rotation2d getRotation2d() {
+        return Rotation2d.fromDegrees(getHeading());
+    }
+
+    @Override
+    public void periodic() {
+        SmartDashboard.putNumber("Robot Heading", getHeading());
+    }
+
+    public void stopModules(){
+        frontLeft.stop();
+        frontRight.stop();
+        backLeft.stop();
+        backRight.stop();
+    }
+
+    public void setModuleStates(SwerveModuleState[] desiredStates) {
+        SwerveDriveKinematics.desaturateWheelSpeeds(desiredStates, DriveConstants.kPhysicalMaxSpeed);
```

Comment:
You do a scaling for x and y speed based on kPhysicalMax Speed.  Why are we also desaturating here?

We need to dig into this further.

Reply:
The code here renormalizes the wheel speeds, in case any individual wheel speed is above the specified maximum.
It's to make sure no module is set to attempt more than it can do.

Reply:
Why are we scaling it in the execute function before this?  Feels like they should be combined in some way.  

We should talk about the code, maybe I am missing something.

Reply:
This code is a separate failsafe to make sure we are _never_ oversaturating wheel speeds, as that would cause a catastrophic, albeit temporary, loss of all turning control. As you said above, there are edge cases that could lead to oversaturation (moving faster diagonally without a speed factor), which we need to protect against.

Reply:
Consensus was that desaturation should be a failsafe, but we should design to never encounter this failsafe, logging a warning if the wheel speeds must be desaturated.